File: csharp/src/BrValidator/Ie/Ie.cs
using System;
using System.Collections.Generic;

namespace BrValidator.Ie
{
    /// <summary>
    /// Validates, normalizes, and formats Brazilian Inscrição Estadual
    /// numbers.
    ///
    /// Inscrição Estadual has no single national rule: each state (SEFAZ)
    /// defines its own digit count and check-digit algorithm, so every UF is
    /// modeled as its own internal class in this namespace and registered
    /// in the States map below. A UF absent from that map is simply not
    /// supported yet.
    /// </summary>
    public static class Ie
    {
        private class StateValidator
        {
            public Func<string, bool> IsValid { get; }
            public Func<string, string> Normalize { get; }
            public Func<string, string> Format { get; }

            public StateValidator(Func<string, bool> isValid, Func<string, string> normalize,
                Func<string, string> format)
            {
                IsValid = isValid;
                Normalize = normalize;
                Format = format;
            }
        }

        private static readonly Dictionary<string, StateValidator> States = new Dictionary<string, StateValidator>
        {
            { "AC", new StateValidator(Ac.IsValid, Ac.Normalize, Ac.Format) },
            { "AL", new StateValidator(Al.IsValid, Al.Normalize, Al.Format) },
            { "AM", new StateValidator(Am.IsValid, Am.Normalize, Am.Format) },
            { "AP", new StateValidator(Ap.IsValid, Ap.Normalize, Ap.Format) },
            { "BA", new StateValidator(Ba.IsValid, Ba.Normalize, Ba.Format) },
            { "CE", new StateValidator(Ce.IsValid, Ce.Normalize, Ce.Format) },
            { "DF", new StateValidator(Df.IsValid, Df.Normalize, Df.Format) },
            { "ES", new StateValidator(Es.IsValid, Es.Normalize, Es.Format) },
            { "GO", new StateValidator(Go.IsValid, Go.Normalize, Go.Format) },
            { "MA", new StateValidator(Ma.IsValid, Ma.Normalize, Ma.Format) },
            { "MG", new StateValidator(Mg.IsValid, Mg.Normalize, Mg.Format) },
            { "MS", new StateValidator(Ms.IsValid, Ms.Normalize, Ms.Format) },
            { "MT", new StateValidator(Mt.IsValid, Mt.Normalize, Mt.Format) },
            { "PA", new StateValidator(Pa.IsValid, Pa.Normalize, Pa.Format) },
            { "PB", new StateValidator(Pb.IsValid, Pb.Normalize, Pb.Format) },
            { "PE", new StateValidator(Pe.IsValid, Pe.Normalize, Pe.Format) },
            { "PI", new StateValidator(Pi.IsValid, Pi.Normalize, Pi.Format) },
            { "PR", new StateValidator(Pr.IsValid, Pr.Normalize, Pr.Format) },
            { "RJ", new StateValidator(Rj.IsValid, Rj.Normalize, Rj.Format) },
            { "RN", new StateValidator(Rn.IsValid, Rn.Normalize, Rn.Format) },
            { "RO", new StateValidator(Ro.IsValid, Ro.Normalize, Ro.Format) },
            { "RR", new StateValidator(Rr.IsValid, Rr.Normalize, Rr.Format) },
            { "RS", new StateValidator(Rs.IsValid, Rs.Normalize, Rs.Format) },
            { "SC", new StateValidator(Sc.IsValid, Sc.Normalize, Sc.Format) },
            { "SE", new StateValidator(Se.IsValid, Se.Normalize, Se.Format) },
            { "SP", new StateValidator(Sp.IsValid, Sp.Normalize, Sp.Format) },
            { "TO", new StateValidator(To.IsValid, To.Normalize, To.Format) },
        };

        private static StateValidator ResolveState(string uf)
        {
            States.TryGetValue(uf.Trim().ToUpperInvariant(), out StateValidator state);
            return state;
        }

        /// <summary>Returns true if value is a valid Inscrição Estadual for uf.</summary>
        public static bool IsValid(string value, string uf)
        {
            StateValidator state = ResolveState(uf);

            return state != null && state.IsValid(value);
        }

        /// <summary>
        /// Strips formatting and returns the canonical representation of
        /// value for uf. If uf is not supported, value is returned
        /// unchanged.
        /// </summary>
        public static string Normalize(string value, string uf)
        {
            StateValidator state = ResolveState(uf);

            return state == null ? value : state.Normalize(value);
        }

        /// <summary>
        /// Returns value in uf's standard human-readable representation. If
        /// uf is not supported or the normalized value has an invalid
        /// length, value is returned unchanged.
        /// </summary>
        public static string Format(string value, string uf)
        {
            StateValidator state = ResolveState(uf);

            return state == null ? value : state.Format(value);
        }
    }
}

File: java/src/main/java/io/github/matheuslm7/brvalidator/ie/Ie.java
package io.github.matheuslm7.brvalidator.ie;

import java.util.Map;
import java.util.function.Predicate;
import java.util.function.UnaryOperator;

/**
 * Validates, normalizes, and formats Brazilian Inscrição Estadual numbers.
 *
 * <p>Inscrição Estadual has no single national rule: each state (SEFAZ) defines its own digit
 * count and check-digit algorithm, so every UF is modeled as its own package-private class in
 * this package and registered in the {@code STATES} map below. A UF absent from that map is
 * simply not supported yet.
 */
public final class Ie {

  private static final class StateValidator {
    final Predicate<String> isValid;
    final UnaryOperator<String> normalize;
    final UnaryOperator<String> format;

    StateValidator(Predicate<String> isValid, UnaryOperator<String> normalize,
        UnaryOperator<String> format) {
      this.isValid = isValid;
      this.normalize = normalize;
      this.format = format;
    }
  }

  private static final Map<String, StateValidator> STATES = Map.ofEntries(
      Map.entry("AC", new StateValidator(Ac::isValid, Ac::normalize, Ac::format)),
      Map.entry("AL", new StateValidator(Al::isValid, Al::normalize, Al::format)),
      Map.entry("AM", new StateValidator(Am::isValid, Am::normalize, Am::format)),
      Map.entry("AP", new StateValidator(Ap::isValid, Ap::normalize, Ap::format)),
      Map.entry("BA", new StateValidator(Ba::isValid, Ba::normalize, Ba::format)),
      Map.entry("CE", new StateValidator(Ce::isValid, Ce::normalize, Ce::format)),
      Map.entry("DF", new StateValidator(Df::isValid, Df::normalize, Df::format)),
      Map.entry("ES", new StateValidator(Es::isValid, Es::normalize, Es::format)),
      Map.entry("GO", new StateValidator(Go::isValid, Go::normalize, Go::format)),
      Map.entry("MA", new StateValidator(Ma::isValid, Ma::normalize, Ma::format)),
      Map.entry("MG", new StateValidator(Mg::isValid, Mg::normalize, Mg::format)),
      Map.entry("MS", new StateValidator(Ms::isValid, Ms::normalize, Ms::format)),
      Map.entry("MT", new StateValidator(Mt::isValid, Mt::normalize, Mt::format)),
      Map.entry("PA", new StateValidator(Pa::isValid, Pa::normalize, Pa::format)),
      Map.entry("PB", new StateValidator(Pb::isValid, Pb::normalize, Pb::format)),
      Map.entry("PE", new StateValidator(Pe::isValid, Pe::normalize, Pe::format)),
      Map.entry("PI", new StateValidator(Pi::isValid, Pi::normalize, Pi::format)),
      Map.entry("PR", new StateValidator(Pr::isValid, Pr::normalize, Pr::format)),
      Map.entry("RJ", new StateValidator(Rj::isValid, Rj::normalize, Rj::format)),
      Map.entry("RN", new StateValidator(Rn::isValid, Rn::normalize, Rn::format)),
      Map.entry("RO", new StateValidator(Ro::isValid, Ro::normalize, Ro::format)),
      Map.entry("RR", new StateValidator(Rr::isValid, Rr::normalize, Rr::format)),
      Map.entry("RS", new StateValidator(Rs::isValid, Rs::normalize, Rs::format)),
      Map.entry("SC", new StateValidator(Sc::isValid, Sc::normalize, Sc::format)),
      Map.entry("SE", new StateValidator(Se::isValid, Se::normalize, Se::format)),
      Map.entry("SP", new StateValidator(Sp::isValid, Sp::normalize, Sp::format)),
      Map.entry("TO", new StateValidator(To::isValid, To::normalize, To::format))
  );

  private Ie() {}

  private static StateValidator resolveState(String uf) {
    return STATES.get(uf.trim().toUpperCase());
  }

  /** Returns true if value is a valid Inscrição Estadual for uf. */
  public static boolean isValid(String value, String uf) {
    StateValidator state = resolveState(uf);

    if (state == null) {
      return false;
    }

    return state.isValid.test(value);
  }

  /**
   * Strips formatting and returns the canonical representation of value for uf.
   *
   * <p>If uf is not supported, value is returned unchanged.
   */
  public static String normalize(String value, String uf) {
    StateValidator state = resolveState(uf);

    if (state == null) {
      return value;
    }

    return state.normalize.apply(value);
  }

  /**
   * Returns value in uf's standard human-readable representation.
   *
   * <p>If uf is not supported or the normalized value has an invalid length, value is returned
   * unchanged.
   */
  public static String format(String value, String uf) {
    StateValidator state = resolveState(uf);

    if (state == null) {
      return value;
    }

    return state.format.apply(value);
  }
}
